speedtest/test_performance_openssl.py

```diff
@@ -0,0 +1,247 @@
+#!/usr/bin/python
+"""
+A py.test test that attempts to build openssl and benchmark the effect of clcache
+"""
+import sys
+import os
+import shutil
+import pytest
+import urllib
+import zipfile
+import subprocess
+import time
+
+
+OPENSSL_ZIP = "OpenSSL_1_0_2-stable.zip"
+OPENSSL_URL = "https://codeload.github.com/openssl/openssl/zip/OpenSSL_1_0_2-stable"
+
+THISDIR = os.path.dirname(os.path.abspath(__file__))
+DISTDIR = os.path.join(os.path.dirname(THISDIR), "dist")
+CLCACHE = os.path.join(DISTDIR, "cl.exe")
+SOURCES = []
+ENVS = dict(os.environ)
+os.chdir(THISDIR)
+
+
+def retry_delete(path):
+    """
+    Repeatedly attempt to delete path
+    :param path:
+    :return:
+    """
+    for _ in range(30):
+        # antivirus might be busy in here..
+        try:
+            shutil.rmtree(path)
+            return
+        except WindowsError:
+            time.sleep(2)
+        if os.path.exists(path):
+            raise Exception("could not delete {}".format(path))
+
+
+class BlockMessage(object):
+    """
+    Little class to emit "begin .. end" messages for a block of code
+    """
+
+    def __init__(self, message):
+        self.started = 0
+        self.message = message
+
+    def __enter__(self):
+        self.started = time.time()
+        print "\n..begin {} .. ".format(self.message)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        result = "OK"
+        if exc_val is not None:
+            result = "ERROR"
+        print "\n..end {} {}.. ({}sec)".format(self.message, result,
+                                               time.time() - self.started)
+
+
+def download_openssl():
+    """
+    Get the openssl zip and unpack it
+    """
+    if not os.path.exists(OPENSSL_ZIP):
+        getoslsrc = urllib.URLopener()
+        with BlockMessage("download openssl"):
+            getoslsrc.retrieve(OPENSSL_URL, OPENSSL_ZIP + ".part")
+            os.rename(OPENSSL_ZIP + ".part", OPENSSL_ZIP)
+
+
+def clean_openssl_build():
+    """
+    Unpack the openssl source, possibly deleting the previous one
+    :return:
+    """
+    with zipfile.ZipFile(OPENSSL_ZIP, "r") as unzip:
+        folder = unzip.namelist()[0]
+        if os.path.exists(folder):
+            with BlockMessage("delete old openssl folder"):
+                retry_delete(folder)
+
+        with BlockMessage("unzip openssl"):
+            unzip.extractall()
+
+        if len(SOURCES) == 0:
```

Comment:
Regarding testing sequences for emptyness, [PEP 8|https://www.python.org/dev/peps/pep-0008/] says:

> For sequences, (strings, lists, tuples), use the fact that empty sequences are false.
>
> Yes: if not seq:
>      if seq:
>
> No: if len(seq):
>     if not len(seq):

Based on that, I think `if not SOURCES:` would be better here.